src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/ResetUserPasswordTask.cs

```diff
@@ -0,0 +1,159 @@
+        }
+
+        // This is the heart of the activity and actually performs the work to be done.
+        public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
+        {
+            var userName = await _expressionEvaluator.EvaluateAsync(UserName, workflowContext, null);
+
+            var user = await _userService.GetForgotPasswordUserAsync(userName) as User;
+            if (user == null || ((await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>().UsersMustValidateEmail && !await _userManager.IsEmailConfirmedAsync(user)))
+            {
+                var updater = _updateModelAccessor.ModelUpdater;
+                if (updater != null)
+                {
+                    updater.ModelState.TryAddModelError("", S["No email service is available"]);
+                }
+                return Outcomes("Invalid");
+            }
+            else
+            {
+                user.ResetToken = Convert.ToBase64String(Encoding.UTF8.GetBytes(user.ResetToken));
+                var resetPasswordUrl = _linkGenerator.GetUriByAction(_httpContextAccessor.HttpContext,
+                    "ResetPassword", "ResetPassword", new { area = "OrchardCore.Users", code = user.ResetToken });
+
+                workflowContext.Properties["ResetPasswordUrl"] = resetPasswordUrl;
+
+                var subject = await _expressionEvaluator.EvaluateAsync(ResetPasswordEmailSubject, workflowContext, null);
+                var localizedSubject = new LocalizedString(nameof(ResetUserPasswordTask), subject);
+
+                var body = await _expressionEvaluator.EvaluateAsync(ResetPasswordEmailTemplate, workflowContext, _htmlEncoder);
+                var localizedBody = new LocalizedHtmlString(nameof(ResetUserPasswordTask), body);
+                var message = new MailMessage()
+                {
+                    To = user.Email,
+                    Subject = localizedSubject.ResourceNotFound ? subject : localizedSubject.Value,
+                    Body = localizedBody.IsResourceNotFound ? body : localizedBody.Value,
+                    IsBodyHtml = true
+                };
+                var smtpService = _httpContextAccessor.HttpContext.RequestServices.GetService<ISmtpService>();
+
+                if (smtpService == null)
+                {
+                    var updater = _updateModelAccessor.ModelUpdater;
+                    if (updater != null)
+                    {
+                        updater.ModelState.TryAddModelError("", S["No email service is available"]);
+                    }
+                    return Outcomes("Invalid");
+                }
+                else
+                {
+                    var result = await smtpService.SendAsync(message);
+                    if (!result.Succeeded)
+                    {
+                        var updater = _updateModelAccessor.ModelUpdater;
```

Comment:
Why is the ModelUpdater used at this point? Or anywhere else in the Execute method?

Reply:
These same lines also exist in RegisterUserTask.cs as well. Is there a specific reason for them here?

Reply:
Please change it to common log. The other task will have to be fixed too.

Reply:
@sebastienros @JoshLefebvre 

I don't remember all details but e.g. the `ValidateFormFieldTask`, that may be triggered by a form submission, uses data from the `HttpContext.Request.Form` and then **uses the underlying `ModelUpdater`** so that validation error messages are shown. The `RegisterUserTask` also uses data from the `HttpContext.Request.Form` and then **uses the underlying `ModelUpdater`** for the same reasons.


Here it seems that all data come from the activity properties **but i would say it depends in which context this activity may be executed**. E.g. in #6054 i also use the underlying model updater if we detect that we are executing inline after having edited a content item and then submitted it. Otherwise e.g. the `UpdateContentTask` may fail to validate the content item and cancel the session, **but you would still see a message saying e.g. that it has been saved**.

Note: For infos, currently if the validation fails in the `UpdateContentTask` we don't log the errors but we pass them in the WF `LastResult` so that it may be used by a subsequent activity e.g. a `NotifyTask`.

    workflowContext.LastResult = result; // a ContentValidateResult containing the errors

    return Outcomes("Failed");